vision_transformer.py

```diff
@@ -0,0 +1,203 @@
+                     out_feat_keys]
+        n_blk_save = max([n for name, n in out_feats if name == "BLK"] + [0])
+        n_cls_save = max([n for name, n in out_feats if name == "CLS"] + [0])
+
+        after_i_blocks_save_blk = len(self.blocks) - n_blk_save + 1
+        after_i_blocks_save_cls = len(self.blocks) - n_cls_save + 1
+        after_i_blocks_save = min(after_i_blocks_save_blk, after_i_blocks_save_cls)
+        assert (after_i_blocks_save >= K)
+
+        # run separately
+        subencoder = nn.Sequential(*self.blocks[:K])
+        xs = [subencoder(x) for x in xs]
+
+        # mixing
+        xs_cls = torch.stack([x[:, [0], :] for x in xs])
+        xs_feats = [x[:, 1:, :] for x in xs]
+        x = torch.cat([xs_cls.mean(dim=0)] + xs_feats, dim=1)
+
+        for blk in self.blocks[K:after_i_blocks_save]:
+            x = blk(x)
+
+        # extract
+        blk_feats = []
+        cls_feats = []
+
+        if after_i_blocks_save >= after_i_blocks_save_blk:
+            blk_feats.append(self.norm(x))
+        if after_i_blocks_save >= after_i_blocks_save_cls:
+            cls_feats.append(self.norm(x[:, 0, :]))
+
+        for i, blk in enumerate(self.blocks[after_i_blocks_save:]):
+            x = blk(x)
+            if i + after_i_blocks_save >= after_i_blocks_save_blk:
+                blk_feats.append(self.norm(x))
+            if i + after_i_blocks_save >= after_i_blocks_save_cls:
+                cls_feats.append(self.norm(x[:, 0, :]))
+
+        if len(out_feats) > 0:
+            output = [
+                torch.cat(cls_feats[-n:], dim=-1) if feat == "CLS" else torch.cat(blk_feats[-n:], dim=-1)
+                for feat, n in out_feats
+            ]
+        else:
+            output = x[:, 0, :]
+        return output
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    @torch.jit.ignore
+    def no_weight_decay(self):
+        return {'pos_embed', 'cls_token'}
+
+    def get_classifier(self):
+        return self.head
+
+    def reset_classifier(self, num_classes, global_pool=''):
+        self.num_classes = num_classes
+        self.head = nn.Linear(self.embed_dim, num_classes) if num_classes > 0 else nn.Identity()
+
+    def forward_features(self, x):
+        B = x.shape[0]
+        x = self.patch_embed(x)
+
+        cls_tokens = self.cls_token.expand(B, -1, -1)  # stole cls_tokens impl from Phil Wang, thanks
+        x = torch.cat((cls_tokens, x), dim=1)
+        x = x + self.pos_embed
+        x = self.pos_drop(x)
+
+        for blk in self.blocks:
+            x = blk(x)
+
+        x = self.norm(x)[:, 0]
+        x = self.pre_logits(x)
+        return x
+
+    def forward(self, sample):
```

Comment:
Change the signature to forward(x, K=0, M=1), forward should accept 1 tensor positional argument

Reply:
Done